cmd/soroban-cli/src/strval.rs

```diff
@@ -29,76 +40,454 @@ impl From<()> for Error {
+                    .map(|item| self.from_json(item, &elem.element_type))
+                    .collect::<Result<Vec<ScVal>, Error>>()?
+                    .try_into()
+                    .map_err(Error::Xdr)?;
+                ScVal::Object(Some(ScObject::Vec(converted)))
             }
-        }
 
-        // Might have wrapping quotes if it is negative. e.g. "-5"
-        ScSpecTypeDef::I128 => {
-            if let Ok(Value::String(raw)) = serde_json::from_str(s) {
-                // First, see if it is a json string, strip the quotes and recurse
-                from_string(&raw, t)?
-            } else {
-                i128::from_str(s)
-                    .map_err(|_| Error::InvalidValue(Some(t.clone())))?
-                    .into()
+            // Map parsing
+            (ScSpecTypeDef::Map(map), Value::Object(raw)) => self.parse_map(map, raw)?,
+
+            // Option parsing
+            // is null -> void the right thing here?
+            (ScSpecTypeDef::Option(_), Value::Null) => ScVal::Object(None),
+            (ScSpecTypeDef::Option(elem), v) => ScVal::Object(Some(
+                self.from_json(v, &elem.value_type)?
+                    .try_into()
+                    .map_err(|_| Error::InvalidValue(Some(t.clone())))?,
+            )),
+
+            // Tuple parsing
+            (ScSpecTypeDef::Tuple(elem), Value::Array(raw)) => self.parse_tuple(t, elem, raw)?,
+
+            (ScSpecTypeDef::Udt(ScSpecTypeUdt { name }), _) => self.parse_udt(name, v)?,
+
+            // TODO: Implement the rest of these
+            // ScSpecTypeDef::Bitset => {},
+            // ScSpecTypeDef::Status => {},
+            // ScSpecTypeDef::Result(Box<ScSpecTypeResult>) => {},
+            // ScSpecTypeDef::Set(Box<ScSpecTypeSet>) => {},
+            // ScSpecTypeDef::Udt(ScSpecTypeUdt) => {},
+            (_, raw) => serde_json::from_value(raw.clone()).map_err(Error::Serde)?,
+        };
+        Ok(val)
+    }
+
+    fn parse_udt(&self, name: &StringM<60>, value: &Value) -> Result<ScVal, Error> {
+        let name = &name.to_string_lossy();
+        match (self.find(name)?, value) {
+            (ScSpecEntry::UdtStructV0(strukt), Value::Object(map)) => {
+                self.parse_strukt(strukt, map)
+            }
+            (ScSpecEntry::UdtStructV0(strukt), Value::Array(arr)) => {
+                self.parse_tuple_strukt(strukt, arr)
             }
+            (ScSpecEntry::UdtUnionV0(union), val @ (Value::String(_) | Value::Object(_))) => {
+                self.parse_union(union, val)
+            }
+            (ScSpecEntry::UdtEnumV0(enum_), Value::Number(num)) => parse_const_enum(num, enum_),
+            (s, v) => todo!("Not implemented for {s:#?} {v:#?}"),
         }
+    }
 
-        // For all others we just use the json parser
-        _ => serde_json::from_str(s)
-            .map_err(Error::Serde)
-            .and_then(|raw| from_json(&raw, t))?,
-    };
-    Ok(val)
+    fn parse_tuple_strukt(
```

Comment:
```suggestion
    fn parse_tuple_struct(
```

Reply:
Siem en stuct is a keyword I used strukt throughout. It's what rust analyzer does. But here I guess it doesn't matter as much. Definitely no strong feelings if you want to rename it.

Reply:
Ok. if rust analyzer does it, then fine. Thought it was just a typo. :)